# dynamic_biomarkers
Code and data corresponding to Villafranca et al., 2025

WGCNA FOLDER: This folder has data and code corresponding to all WGCNA analyses (T0, T12, and module preservation). These code were used to obstain data used for Figure 2, and the output files for Figure 3. Note that all trait data was obtained from https://github.com/wyattmillion/Acer_Morphological_Plasticity

GO FOLDER: This folder has data and code corresponding to GO analyses -- note that one GO_MWU output file to be fed into GO_plot_purple could not be attached due to size limitations. This data and code was used to to run statistical analyses and specific gene expression analyses that were the basis for figure 3.

MICROBIOME FOLDER: This folder has data and code corresponding to microbiome analysis in supplemental data. Original data and code can be found at https://github.com/symbiotic-em/acer_epi_final

Figure1_survival_vs_resval.R was used to run statistics on comparison between survival ranking and restoration value ranking.

Figure4_2018_vs_2022.R was used to run analysis comparing 2018 survival data to 2022 survival data.

ReadsTable has the number of counts at each step of QC/QF of TagSeq data prior to WGCNA input. 
